### ISSUE TITLE
Devs: objects/UUID compare problems (inner rollbacks problem)

### ISSUE THREAD

**Issue by JayDi85** [magefree/mage#4407]
Continue from #4402

You can't compare UUID or objects by `==` operand. Need to use only `obj.equals()`.

IntelliJ IDEA have inspections on that problems (use custom inspect settings to filter only needed data):
![shot_180113_143953](https://user-images.githubusercontent.com/8344157/34905603-fe1ffe40-f875-11e7-90de-5853a9bdd061.png)

Current project have 72 warnings with `==` on objects (exported [list to web page here](http://jaydi.ru/xmage/dev-compare/)):
![shot_180113_152710](https://user-images.githubusercontent.com/8344157/34905617-438000a2-f876-11e7-938f-e96a137341be.png)

**BUT some xmage objects do not overrides Equals method and use default one, e.g. compare objects by `==` (compare refs, not data) .** 

As example, `Player` objects do not have Equals. All players data searched from game's `state` object:
![shot_180113_151814](https://user-images.githubusercontent.com/8344157/34906210-1879163a-f882-11e7-9f14-7de9777d30c1.png)

But game `state` can be changed and reverted. In particular, it's saved to history every step by COPY (e.g. new state's players is not equals to old one). And on rollback step can broke something if you save ref object to somewhere (is that can cause of "buggy rollback" function?) or if you have another thread (is xmage use one thread per game?):
![shot_180113_152232](https://user-images.githubusercontent.com/8344157/34906222-5f932d80-f882-11e7-9320-cb0e873a6aba.png)

I don't known confirmed bug with that issue like with static filters (#4402). But devs have to be careful about this.

Howto fix that:
 * Enable IDE warnings, find and replace `==` to `equals` for xmage and UUID objects compare;
 * Add to each needed object overrided `equals` method. Example for `PlayerImp.java` (players with same ID will be equal):
```
@Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        PlayerImpl obj = (PlayerImpl) o;
        if (this.getId() == null || obj.getId() == null) {
            return false;
        }

        return this.getId().equals(obj.getId());
    }
```

**Comment by LevelX2** [comment]
Good findings! 👍

**Comment by ingmargoudt** [comment]
Very good finding!!

**Comment by Zzooouhh** [comment]
Interesting, does it impact OneShotEffects too (since they're presumably reverted along with the rollback), or is only a problem for continuous effects in practice?

**Comment by JayDi85** [comment]
Restore/rollback works dangerous. Looked at code. Game save new state every steps (save a copy of current state with all data, triggers, effects, players, cards, etc). 

On restore request it search needed state by index and delete all other after that state:

![shot_180113_195424](https://user-images.githubusercontent.com/8344157/34907694-e5c47472-f89b-11e7-844a-581d498c240b.png)

And replace all current state's data with old one. Old triggers copy too -- it's can't get current effects:
![shot_180113_200309](https://user-images.githubusercontent.com/8344157/34907760-d183de34-f89c-11e7-9fb3-43849be14b86.png)

If something have ref to old state's data on that point then it will be broken (like another thread or code execute after restore point).

P.S. O_o Found that restore/rollback used in targets process -- if was wrong target then all state restored O_o It's may be very buggy and broke the game even without rollbacks (if that "rollback" doesn't work correct):
![shot_180113_201249](https://user-images.githubusercontent.com/8344157/34907842-558dbb5e-f89e-11e7-947d-498261515796.png)

**Comment by JayDi85** [comment]
@Zzooouhh yes, replacement bugs from restore/rollback function. That's it. It use save state before cast and then restored after cast (e.g. all data checks after restore will be dangerous to use by refs):

![shot_180113_201808](https://user-images.githubusercontent.com/8344157/34907885-2c1f8094-f89f-11e7-86ef-98460eae9874.png)

![shot_180113_201830](https://user-images.githubusercontent.com/8344157/34907886-30337a00-f89f-11e7-8ce6-e556b850cc7e.png)

**Comment by JayDi85** [comment]
Another bad thing (I think) -- on wrong ability it's try to restore AND after restore use old objects/ref to old state. But that old state was removed and inaccessible. Something bad may happing here (or may not -- I don't know details of java VM and GC works).

![shot_180113_202611](https://user-images.githubusercontent.com/8344157/34907921-0d7ccbb4-f8a0-11e7-8597-9d02efb3d423.png)

**Comment by JayDi85** [comment]
As a fast workaround it may be good idea do not remove all old state on restore, but moved it to temp object (e.g. ensure that all old refs will works to the end). Maybe debugger or JVM have feature to find out all refs to that object after it was moved to temp -- that's can help to find wrong code.

**Comment by LevelX2** [comment]
Normally Players or MageObjects won't be checked directly. 
They are compared by their UUID and if equals is used for the ID it works as intended.

So I don't see that bugy anywhere. But indeed we have to replace == checks on the UUIDs with equals.

**Comment by JayDi85** [comment]
Well, I've checked and fixed warning from IDE about `==` problems in c24ba742f685218750dd7231dc5701557dc72a89. It's can be closed. But devs must check static analyser warnings time by time.

**Comment by JayDi85** [comment]
Found real life proof (#5784) of that problem:

![shot_190506_001520](https://user-images.githubusercontent.com/8344157/57200014-1daafa00-6f97-11e9-82a0-412482429c5e.png)

**Comment by JayDi85** [comment]
Another one example of rollback and game state problem (#5835 with `buyback` ability -- a few bugs have allowed this ability to work):
 * xmage uses copies of spells/abilities to play in many places, not original objects;
![shot_190616_214128](https://user-images.githubusercontent.com/8344157/59567485-864dc200-907f-11e9-801b-0332b83b8a2a.png)
 * some objects can store inner data/statuses;
![shot_190616_213823](https://user-images.githubusercontent.com/8344157/59567465-2b1bcf80-907f-11e9-8322-79419c01aa30.png)
 * some objects forget to make copy and uses one object ref in all instances (e.g. same object in all game states):
![shot_190616_214606](https://user-images.githubusercontent.com/8344157/59567542-299ed700-9080-11e9-805c-75f2450fdc93.png)
* some objects can change inner data and that's data can be used outside of that objects:
![shot_190616_215033](https://user-images.githubusercontent.com/8344157/59567583-ca8d9200-9080-11e9-9a84-748051b61bbc.png)
![shot_190616_215117](https://user-images.githubusercontent.com/8344157/59567602-e5f89d00-9080-11e9-96e9-4c667b2d7829.png)
* that's wrong -- you must use game state to store dynamic and shared data not inner objects:
![shot_190616_215210](https://user-images.githubusercontent.com/8344157/59567623-0de80080-9081-11e9-8c9e-6d1f9ec4c388.png)
